### PR TITLE
[JIT][NNC] Add handling of strides to dynamic shape support.

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71173
* #71138
* #70656
* __->__ #70464
* #71144

Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:

```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

### CODE REVIEW

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 1]
```diff
@@ -1171,7 +1171,7 @@ class TensorExprFuser {
     }
     for (Node* fusion_group : fusion_groups) {
       VLOG(1) << "GenerateGuard for fusion group: " << *fusion_group;
-      if (!GenerateGuard(fusion_group, /*add_composed_op=*/true)) {
+      if (!GenerateGuard(fusion_group, /*add_composed_op=*/false)) {
```
Comment (eellison) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:1174]: TODO: make this `/*add_composed_op*/` into a flag for fusion, cc @navahgar

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 2]
```diff
@@ -1416,7 +1501,8 @@ std::vector<CodeGen::CallArg> TensorExprKernel::prepareRunArgs(
   // TODO: preallocate `runArgs` during compilation and fill in values where
   // possible (e.g. for constant tensors)
   std::vector<CodeGen::CallArg> runArgs;
-  runArgs.reserve(inputs.size() + bufOutputs_.size());
+  runArgs.reserve(
+      inputs.size() + input_stride_args_.size() + bufOutputs_.size());
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1505]: If the strides here are runtime args, shouldn't that be already part of `inputs`?

File: aten/src/ATen/core/jit_type.h [hunk 3]
```diff
@@ -26,6 +26,23 @@ using OptNameList = c10::optional<std::vector<std::string>>;
 void standardizeVectorForUnion(std::vector<TypePtr>& reference, std::vector<TypePtr>* to_fill);
 void standardizeVectorForUnion(std::vector<TypePtr>* to_flatten);
 
+inline bool is_contiguous_strides(
+    const IntArrayRef sizes,
+    const IntArrayRef strides) {
+  int n_dim = static_cast<int>(sizes.size());
+
+  if (n_dim == 0 || strides[n_dim - 1] != 1) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply aten/src/ATen/core/jit_type.h:34]: yea, this was just factoring out this function, I didn't make any changes, but good point I'll update

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp [hunk 4]
```diff
@@ -69,25 +69,124 @@ std::map<int64_t, Value*> InsertSymbolicShapesCompute(
...
+        summarizeStrideDim(sizes, strides, dim, stride_inputs, 0));
+  }
+  return stride_inputs;
+};
+
+// Todo: incorporate in codegen
+StrideInput summarizeOutputStrides(const TensorType& tt) {
+  auto strides = *tt.strides().concrete_sizes();
+  auto sizes = *tt.sizes().concrete_sizes();
+  // We only try to maintain output striding for channels last tensors,
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:157]: What do you mean ?  In other cases, we will just make it contiguous

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp [hunk 5]
```diff
@@ -309,49 +470,70 @@ RegisterOperators reg_guard({
...
             std::vector<int64_t> flattened_symbolic_dims(num_symbolic_dims, -1);
             size_t flattened_dim_offset = 0;
+            size_t flattened_stride_offset = 0;
             for (const auto i : c10::irange(num_inputs)) {
               at::Tensor tensor = inputs[i].toTensor();
               if (C10_UNLIKELY(
                       tensor.device() != device ||
                       tensor.dtype() != expected_scalar_types[i]) ||
                   tensor.requires_grad()) {
-                std::cout << " DEVICE /DTYPE/ REQUIRES GRAD MISMATCH\n";
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:320]: I'm wary of runtime VLOG's, because this will get executed on every execution. Even if debug is set to false, we have to make a call to look up the environment flag which is slow

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.h [hunk 6]
```diff
@@ -33,5 +34,24 @@ TORCH_API void runTensorExprDynamicGroup(
...
+  // than looping through the sizes/strides yourself
+  // For tensors with these properties, we only store one value:
+  TENSOR_CONT,
+  TENSOR_CONT_CHANNELS_LAST,
+  // now, we describe other cases, where there is one stride enum
+  // per dimension
+  S_ONE, // STRIDE_ONE: packed
+  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
+  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
+  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.h:50]: If it's arbitrary it becomes `S_AS_ARG`

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1505]: they're not arguments to the graph, they're taken from the input tensors themselves

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp [hunk 7]
```diff
@@ -69,25 +69,124 @@ std::map<int64_t, Value*> InsertSymbolicShapesCompute(
...
+  } else if (
+      dim + 1 < sizes.size() &&
+      strides[dim] == strides[dim + 1] * sizes[dim + 1]) {
+    return StrideInput::S_CONT;
+    // Transposed Contiguous depends on prior dim and contiguous depends on next
+    // dim, so to avoid a mutual dependence check that the next dim is Stride
+    // Contiguous
+  } else if (
+      dim > 0 && strides[dim] == strides[dim - 1] * sizes[dim - 1] &&
+      (stride_inputs[dim - 1 + stride_inputs_offset] != StrideInput::S_CONT)) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:132]: The previous StrideInput may not exist if `dim ==0`, and we would be incurring an extra read of a value that may not be used

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

**Comment by eellison** [comment]
With contiguous inputs
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(1, SS(-6), requires_grad=0, device=cpu)], striding_inputs_desc=[["TENSOR_CONT"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=1]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(1, 10, strides=[10, 1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-6], striding_inputs_desc=[["TENSOR_CONT"]], striding_outputs_desc=[["TENSOR_CONT"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
with prim::TensorExprGroup_0 = graph(%x : Double(1, SS(-6), requires_grad=0, device=cpu),
      %SS_6 : int):
  %3 : Double(1, SS(-6), strides=[10, 1], requires_grad=0, device=cpu) = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
  %1 : Double(1, SS(-6), requires_grad=0, device=cpu) = aten::sigmoid(%3) # test/test_jit_fuser_te.py:2027:0
  return (%1)
```
Transposed arg and 1
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-8), SS(-8), requires_grad=0, device=cpu)], striding_inputs_desc=[["S_ONE", "S_TRAN_CONT"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(10, 10, strides=[1, 10], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-8], striding_inputs_desc=[["S_ONE", "S_TRAN_CONT"]], striding_outputs_desc=[["S_ONE", "S_TRAN_CONT"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
```
As arg input
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-9), requires_grad=0, device=cpu)], striding_inputs_desc=[["S_AS_ARG"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(10, strides=[1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-9], striding_inputs_desc=[["S_AS_ARG"]], striding_outputs_desc=[["S_AS_ARG"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
```
Channels last 
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-11), SS(-12), SS(-13), SS(-14), requires_grad=0, device=cpu)], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = prim::Constant[value=1]()
      %20 : int = prim::Constant[value=2]()
      %21 : int = prim::Constant[value=3]()
      %22 : int = aten::__getitem__(%17, %18)
      %23 : int = aten::__getitem__(%17, %19)
      %24 : int = aten::__getitem__(%17, %20)
      %25 : int = aten::__getitem__(%17, %21)
      %9 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-14, -13, -12, -11], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]], striding_outputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]]](%x, %25, %24, %23, %22)
      -> (%9)
    block1():
      %29 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %30 : Tensor = aten::sigmoid(%29) # test/test_jit_fuser_te.py:2027:0
      %31 : (Tensor) = prim::TupleConstruct(%30)
      %28 : Tensor = prim::TupleUnpack(%31)
      -> (%28)
  return (%13)
```

**Review by navahgar — APPROVED** [review]
Mostly LGTM. Couple of minor comments below.

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:135]: imo that wouldn't be clearer to follow than the logic that exists currently

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:163]: we could add support for permuting to other ordering of dims but other than that would require more substantial changes

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (navahgar) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:163]: We might have to support more generic striding because Hao mentioned that they come up in SR models.

Anyways, the point I had here was just to assert in other cases, so that we are doing something incorrectly. If we can't figure out the other cases here, then no problem.

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:163]: generic striding is already handle for inputs.. it's just that we don't preserve striding for outputs if they aren't contiguous or channels-last

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).
